doc/HOWTO-KAFKA.md

```diff
@@ -0,0 +1,226 @@
+# Kafka exfiltration
+
+In the "daemon mode", Sonar stays memory-resident and pushes data to a network sink; one of these
+sinks is a Kafka broker.  See HOWTO-DAEMON.md for general information about this mode and the
+options available for configuring the Kafka producer in Sonar.
+
+Data and control messages are as described in HOWTO-DAEMON.md, with "topic", "key" and "value"
+having their standard Kafka meanings.
+
+## CONFIGURING A STANDARD KAFKA BROKER
+
+### Topics
+
+For each cluster with canonical name `<cluster>` that is going to be handled by the broker, the broker needs
+to be able to handle messages with these topics coming from the cluster:
+
+```
+<cluster>.sample
+<cluster>.sysinfo
+<cluster>.job
+<cluster>.cluster
+```
+
+The broker also needs to be able to handle these control topics (tentative, may change) that are
+sent from the back-end to the clients in the cluster:
+
+```
+<cluster>.control.node
+<cluster>.control.master
+```
+
+### Testing with Apache Kafka
+
+Test notes with standard Kafka server, see https://kafka.apache.org/quickstart.
+
+#### Setup
+
+You're going to be running several shells, let's call them Zookeeper, Server, Consumer, Work, and Sonar.
+
+The working directory for the following is the root directory of the unpacked Kafka distribution, eg
+`kafka_2.13-3.9.0/`.
+
+NOTE!  Currently Sonar does not work with Kafka 4.0.0, there's a protocol error, not debugged.  Use
+3.9.0 for the time being.
+
+#### 2.13-3.9.0
+
+In the Zookeeper shell:
+
+```
+   bin/zookeeper-server-start.sh config/zookeeper.properties
+```
+
+In the Server shell:
+
+```
+  bin/kafka-server-start.sh config/server.properties
+```
+
+In the Work shell, topics need to be added with `kafka-topics.sh` a la this, if you haven't done it
+before (or if you did it, but did not shut down the broker properly):
+
+```
+  bin/kafka-topics.sh --create --topic test-cluster.hpc.uio.no.sample --bootstrap-server localhost:9092
+```
+
+The topics to add are these (the last two are for control messages):
+
+```
+  test-cluster.hpc.uio.no.sample
+  test-cluster.hpc.uio.no.sysinfo
+  test-cluster.hpc.uio.no.job
+  test-cluster.hpc.uio.no.cluster
+  test-cluster.hpc.uio.no.control.node
+  test-cluster.hpc.uio.no.control.master
+```
+
+#### Running sonar and examining the data
+
+Then from the Sonar root directory, after building it, run Sonar in daemon mode with a suitable
+config file in the Sonar shell:
+
+```
+  target/debug/sonar daemon <config-file>
+```
+
+Sonar will run continuously and start pumping data to Kafka.
+
+For easy testing, run this in the Consumer shell to listen for `sysinfo` messages and echo them:
+
+```
+  bin/kafka-console-consumer.sh --topic 'test-cluster.hpc.uio.no.sysinfo' --bootstrap-server localhost:XXXX
```

Comment:
As above, and also below, move the bootstrap-server argument and break the lines.